### ISSUE TITLE
Prettier fails with templated file names

### ISSUE THREAD

**Issue by yajo** [prettier/prettier#12143]
**Environments:**

- Prettier Version: 2.5.1
- Running Prettier via: CLI 
- Runtime: nodejs v16.13.2
- Operating System: Linux
- Prettier plugins (if any): 

**Steps to reproduce:**

I'm developing a copier template. These support templated file names. The template repo itself has pre-commit with a hook that makes use of prettier. The selected file should simply be ignored by prettier, but in any case prettier fails to find the file; not to process its contents:

```sh
> prettier 'src/{% if ansible %}.ansible-lint{% endif %}' 
[error] Unable to resolve file: src/{% if ansible %}.ansible-lint{% endif %}
[error] ENOTDIR: not a directory, scandir '/var/home/yajo/mydevel/autopretty/src/{% if ansible %}.ansible-lint{% endif %}'
```

**Expected behavior:**
Prettier should be able to find the file, because it exists.

**Actual behavior:**
Prettier fails badly. Same as the now stale https://github.com/prettier/prettier/issues/9430.

**Comment by alexander-akait** [comment]
Because it is glob, you should escape some characters, they have different meaning when you have `{`/`}` and other special characters

**Comment by Yajo** [comment]
The escaping needs to be done AFAIK in the shell itself, where it's not needed when it's inside single quotes:

```sh
> ls 'src/{% if ansible %}.ansible-lint{% endif %}'
'src/{% if ansible %}.ansible-lint{% endif %}'
```

**Comment by duailibe** [comment]
> All other tools work perfectly with the same exact parameter

That's not true:

```
> cat 'src/{% if ansible %}.ansible-lint{% endif %}'
cat: src/{% if ansible %}.ansible-lint{% endif %}: No such file or directory
```

Prettier expects valid filenames and fails if an invalid filename is given. I don't think we should change anything here.

**Comment by Yajo** [comment]
Did you create that file before testing? I didn't specify that step, but obviously you need to create a file with that name before passing it to prettier.

**Comment by duailibe** [comment]
Ah, apologies. I didn't understand that before

We use [fast-glob](https://www.npmjs.com/package/fast-glob) for finding files, so there's not much we can do in Prettier.

Also you should be aware that when the file has no extension and Prettier can't infer a parser, it will fail as well:

```
> prettier .ansible-lint 
.ansible-lint[error] No parser could be inferred for file: .ansible-lint
```

**Comment by Yajo** [comment]
Ah, so you mean that prettier parses incoming file names as file patterns always? That's weird. How are we supposed to use it when the file name contains something that looks like a pattern?

In most apps, the shell itself solves the patterns and passes those to the executable being called. Can't we have something like that in prettier? IMHO that should be the default behavior.



> Also you should be aware that when the file has no extension and Prettier can't infer a parser, it will fail as well

Not with `--ignore-unknown`, which is used in the pre-commit hook to allow flexibility in installing plug-ins: https://github.com/pre-commit/mirrors-prettier/blob/ea782651a7e32f40a3d13b76c79d5a2474ee8723/.pre-commit-hooks.yaml#L3

**Comment by duailibe** [comment]
> Ah, so you mean that prettier parses incoming file names as file patterns always?

That's not what I meant, but I think I didn't setup the environment the same way you mentioned, let me investigate a bit more

> Not with --ignore-unknown, which is used in the pre-commit hook to allow flexibility in installing plug-ins

Yes, that's correct.

**Comment by duailibe** [comment]
> Ah, so you mean that prettier parses incoming file names as file patterns always? That's weird. How are we supposed to use it when the file name contains something that looks like a pattern?

No, we do check if the pattern passed happens to match a filename in the filesystem. But in that case we pass it on to fast-glob which then considers it as a pattern, and fails the CLI.

https://github.com/prettier/prettier/blob/a0cc4ab43ba1920e37bb685e1651b36d2e1fda2e/src/cli/expand-patterns.js#L73-L77

**Comment by duailibe** [comment]
What I've figured out so far is that fast-glob fails to match the file if the filename ends with `{...}` (and an "empty" `{}` as well), if `{` and `}` are escaped.

If it has `{}` in the middle of a file (e.g. `src/{{ var }}.js`) or if the brackets aren't escaped it works correctly.

And the failure here is ENOTDIR.

Will continue to investigate.

**Comment by duailibe** [comment]
Tracked it down to the error in fast-glob. Once a fix is released there, we can fix it here.

**Comment by yajo** [comment]
Any progress here? What's the upstream issue in fast-glob?
